sql/catalyst/src/main/scala/org/apache/spark/sql/types/StructType.scala

```diff
@@ -104,6 +104,67 @@ case class StructType(fields: Array[StructField]) extends DataType with Seq[Stru
   private lazy val nameToIndex: Map[String, Int] = fieldNames.zipWithIndex.toMap
 
   /**
+   * Creates a new [[StructType]] by adding a new field.
+   * {{{
+   * val struct = (new StructType)
+   *  .add(StructField("a", IntegerType, true))
+   *  .add(StructField("b", LongType, false))
+   *  .add(StructField("c", StringType, true))
+   *}}}
+   */
+  def add(field: StructField): StructType = {
+    StructType(fields :+ field)
+  }
+
+  /**
+   * Creates a new [[StructType]] by adding a new field.
+   *
+   * val struct = (new StructType)
+   *  .add(StructField("a", IntegerType, true))
+   *  .add(StructField("b", LongType, false))
+   *  .add(StructField("c", StringType, true))
+   */
+  def add(name: String, dataType: DataType): StructType = {
+    StructType(fields :+ new StructField(name, dataType, true, Metadata.empty))
+  }
+
+  def add(name: String, dataType: DataType, nullable: Boolean): StructType = {
+    StructType(fields :+ new StructField(name, dataType, nullable, Metadata.empty))
+  }
+
+  def add(name: String,
+    dataType: DataType,
+    nullable: Boolean,
+    metadata: Metadata): StructType = {
```

Comment:
BTW, are we intentionally avoiding default argument values for Java API friendliness here?